codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/parse/JsonParserGenerator.kt

```diff
@@ -0,0 +1,457 @@
+                            }
+                        }
+                    }
+                    rust("return Ok(Some(items))")
+                }
+            }
+        }
+        rust("#T(tokens)?", parser)
+    }
+
+    private fun RustWriter.deserializeMap(shape: MapShape) {
+        val keyTarget = model.expectShape(shape.key.target) as StringShape
+        val fnName = symbolProvider.deserializeFunctionName(shape)
+        val parser = RuntimeType.forInlineFun(fnName, "json_deser") {
+            it.rustBlockTemplate(
+                """
+                pub fn $fnName<'a, I>(tokens: &mut #{Peekable}<I>) -> Result<Option<#{Shape}>, #{Error}>
+                    where I: Iterator<Item = Result<#{Token}<'a>, #{Error}>>
+                """,
+                "Shape" to symbolProvider.toSymbol(shape),
+                *codegenScope,
+            ) {
+                startObjectOrNull {
+                    rust("let mut map = #T::new();", RustType.HashMap.RuntimeType)
+                    objectKeyLoop {
+                        withBlock("map.insert(", ");") {
+                            if (keyTarget.hasTrait<EnumTrait>()) {
+                                rust(
+                                    "#T::from(key.as_escaped_str())",
+                                    symbolProvider.toSymbol(keyTarget)
+                                )
+                            } else {
+                                rust("key.to_unescaped()?.into_owned()")
+                            }
+                            rust(",")
+                            deserializeMember(shape.value, forceOptional = false)
+                        }
+                    }
+                    rust("Ok(Some(map))")
+                }
+            }
+        }
+        rust("#T(tokens)?", parser)
+    }
+
+    private fun RustWriter.deserializeStruct(shape: StructureShape) {
+        val fnName = symbolProvider.deserializeFunctionName(shape)
+        val symbol = symbolProvider.toSymbol(shape)
+        val nestedParser = RuntimeType.forInlineFun(fnName, "json_deser") {
+            it.rustBlockTemplate(
+                """
+                pub fn $fnName<'a, I>(tokens: &mut #{Peekable}<I>) -> Result<Option<#{Shape}>, #{Error}>
+                    where I: Iterator<Item = Result<#{Token}<'a>, #{Error}>>
+                """,
+                "Shape" to symbol,
+                *codegenScope,
+            ) {
+                startObjectOrNull {
+                    Attribute.AllowUnusedMut.render(this)
+                    rustTemplate("let mut builder = #{Shape}::builder();", *codegenScope, "Shape" to symbol)
+                    deserializeStructInner(shape)
+                    withBlock("return Ok(Some(builder.build()", "))") {
+                        if (StructureGenerator.fallibleBuilder(shape, symbolProvider)) {
+                            rustTemplate(".map_err(|_| #{Error}::custom(\"missing field\"))?", *codegenScope)
```

Comment:
don't throw out this error :-) it will tell you what field is missing, and it's a structured error that actually identifies the exact field